Chainlit/literalai-python
# LlamaIndex instrumentation

### Changes:
- added support to `llm.chat` and `llm.stream_chat`, `llm.predict`, `llm.predict_and_call`, `OpenAIAgent.chat` and `FunctionCallingAgent.chat`. 

### To test:
- validate that a call to the following methods creates a single `Generation`
  - `llm.chat`
  - `llm.stream_chat`
  - `llm.predict`

For the specific case of `llm.predict_and_call`, LlamaIndex does not fire an event for the tool calls that take place. Hence:
- either you decorate your tools with Literal AI step tools, in which case you should also wrap the call to `llm.predict_and_call` in a step or a thread (so that the tool steps have a parent, they cannot be standalone)
- or you do not decorate the tools and the `llm.predict_and_call` simply shows a `Generation` log

Ticket for LlamaIndex to add `ToolCallStartEvent` and `ToolCallEndEvent`: https://github.com/run-llama/llama_index/issues/16231

- validate that a call to query engine shows the same thread as we used to have (user message + RAG agent (retrieval + embedding + synthesise) + assistant message)
- validate that a call to `agent.chat` shows:

<img width="433" alt="image" src="https://github.com/user-attachments/assets/8878fec5-6962-4a0b-81af-11ea843d9a13">

Again, tool calls don't show. However, in the specific case of agents, LlamaIndex fires a `AgentToolCallEvent` prior to calling the tool. It's not enough to create a full tool step, and I decided it was best to let users decorate their own FunctionTool if they needed logs at that level. 

There is an option to refactor the span handler logic to catch the FunctionTool spans that are fired. There are a couple issues related to duplicate firing (nested) and a small revamp of the logic to allow the span handler to open/end steps.
